internal/semaphore/semaphore.go

```diff
@@ -0,0 +1,70 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package semaphore
+
+import (
+	"log"
+)
+
+type Semaphore interface {
+	//Acquire a slot in the semaphore
+	Acquire()
+
+	//Release a slot in the semaphore.
+	Release()
+
+	//GetLimit returns semaphore limit.
+	GetLimit() int
+
+	//GetCount returns current number of occupied slots in semaphore.
+	GetCount() int
+
+	//Done closes down the semaphore
+	Done()
+}
+
+type semaphore struct {
+	slots chan struct{}
+}
+
+func NewSemaphore(limit int) Semaphore {
+	if limit < 0 {
+		log.Panic("E! Semaphore size limit must not be negative")
+	}
+
+	sem := &semaphore{
+		slots: make(chan struct{}, limit),
+	}
+
+	for i := 0; i < cap(sem.slots); i++ {
+		sem.slots <- struct{}{}
+	}
+	
+	return sem
+}
+
+func (sem *semaphore) Acquire() {
+	<-sem.slots
+}
```

Comment:
So if the channel is empty, this blocks until something is added to the channel, which makes sense as a semaphore implementation, but consider how we're using it here. This still aligns with what I said before where we are changing this from being debug logging into becoming a limiter for log files that can be monitored. Is that what you intended for this change?